### PR TITLE
Correct historisation when null values swapped

### PR DESCRIPTION
**Pull request by swissrussdev**
* Historization mechanism based on JSON string
* Generalized function to hash a list of columns
* Additional Test for Null-Swaps
* Correction to isDataFrameData equal which was incorrectly failing identical dataframes

### What changes are included in the pull request?

Link to Issue: #110 

### Why are the changes needed?
See issue.

### CODE REVIEW

File: src/main/scala/io/smartdatalake/util/historization/Historization.scala [hunk 1]
```diff
@@ -145,21 +137,25 @@ private[smartdatalake] object Historization extends SmartDataLakeLogger {
       case (None, None) => newFeedDf.columns.sorted.toSeq
       case (Some(_), Some(_)) => throw new ConfigurationException("historizeWhitelist and historizeBlacklist mustn't be used at the same time.")
     }
-    val newFeedHashed = newFeedDf.withColumn("hash", hashFunc(struct(historizeHashCols.map(col): _*)))
+
+    // Generic column expression to generate a JSON string of the hash fields
+    def colHashExpr(cols: Seq[String]): Column = hash(to_json(struct(cols.map(col): _*)))
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/util/historization/Historization.scala:142]: What is the benefit of using to_json hier? The names of the columns can't change inside a DataFrame. In my opinion this just eats performance... but maybe you have a good reason?
I suggest to used directly "hash(cols.map(col): _*)"

Comment (swissrussdev) [reply src/main/scala/io/smartdatalake/util/historization/Historization.scala:142]: Hi @zzeekk ,

See the example in the issue. The structure of the dataframe doesn't change, but which columns are null can do and therefore the solution using values only can miss version changes. Agreed we increase the payload for the hash function but I think that's a price worth paying for accuracy.

Comment (swissrussdev) [reply src/main/scala/io/smartdatalake/util/historization/Historization.scala:142]: ...maybe the hash is redundant?

**Comment by swissrussdev** [comment]
Removed hashing and reverted to direct comparison of json strings. The logic is that time for hashing and string comparison are both proportional to string length but comparison can shortcut on first difference. Since we only compare rows once, there is no incremental benefit from reusing the hash code. I think the performance issues should be minimised with this approach. Direct struct comparison would have been nice, but doesn't seem possible using different data frames.

**Comment by ursin** [comment]
Hi @swissrussdev and @zzeekk, I agree that the hashing of the json string is an overkill and the direct comparison of the strings should work fine. In the future such changes should be accompanied with some benchmark runs. Without that, we don't really know how big changes  impact performance.

**Pull request merged**